proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,661 @@
+stable public API Currently, those modules and APIs are often underscored or
+specifically annotated. While this approach works it comes with downsides such
+as hiding the APIs in code completion.
+
+## Proposed solution
+
+This proposal introduces a new configuration for packages called _package
+traits_. Package authors can define a set of traits in their `Package.swift`
+that their package offers which provide a way to express conditional compilation
+and optional dependencies. Furthermore, a set of default enabled traits can be
+specified.
+
+```swift
+let package = Package(
+    name: "Example",
+    traits: [
+        "Foo",
+        .trait(
+            name: "Bar",
+            enabledTraits: [ // Other traits that are enabled when this trait is being enabled
+                "Foo",
+            ]
+        )
+        .trait(
+            name: "FooBar",
+            enabledTraits: [
+                "Foo",
+                "Bar",
+            ]
+        ),
+        .default(enabledTraits: ["Foo"]), // Defines all the default enabled traits
+    ],
+    /// ...
+)
+```
+
+When depending on a package the default trait is enabled. However, the enabled
+traits can be customized by passing a set of enabled traits when declaring the
+dependency. When specifying the enabled traits of the dependencies the
+`.default` trait can be passed which will enable the default trait. The below
+example enables the default trait and the additional `SomeTrait` of the package.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        traits: [
+            .default,
+            "SomeTrait"
+        ]
+    ),
+]
+```
+
+To disable all traits including the default trait an empty set can be passed.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        traits: [] // All traits are disabled
+    ),
+]
+```
+
+Another common scenario is to enable a trait of a dependency only when a trait
+of the package is enabled. The below example enables the `SomeOtherTrait` when
+the `Foo` trait of this package is enabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        traits:[
+            .trait("SomeOtherTrait", condition: .when(traits: ["Foo"])),
+        ]
+    ),
+]
+```
+
+Conditional dependencies are specified per target and extend the current
+`condition` syntax which is used for specifying platform dependent dependencies.
+
+```swift
+targets: [
+    .target(
+        name: "SomeTarget",
+        dependencies: [
+            .product(
+                name: "SomeProduct",
+                package: "SomePackage",
+                condition: .when(traits: ["Foo"])
+            ),
+        ]
+    )
+]
+```
+
+Lastly, code can be conditionally compiled by checking if a trait is enabled.
```

Comment:
> checking if a trait is enabled

Can a package check only its own traits, or can it also check a dependency's trait? For example, could my package check if NIO's trait is enabled, and switch behavior based on that?

Reply:
Only the traits of the package containing the module. You can't see into the traits of your dependencies. However, what you are describing here can be easily modelled and I describe this slightly above

> Another common scenario is to enable a trait of a dependency only when a trait
of the package is enabled. The below example enables the `SomeOtherTrait` when
the `Foo` trait of this package is enabled.

The intention here is that you re-expose any configuration from your dependencies in your traits.

Reply:
Sounds good. I was just double checking to make sure we don't need to namespace traits by package name.